File: public/java/014-longest-palindromic-substring.java
/*
 * LeetCode Problem #5: Longest Palindromic Substring
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/longest-palindromic-substring/
 */
import java.util.*;

class Solution {

    // APPROACH 1: Brute Force | O(n³) time | O(1) space
    // EXPLAIN: Check every substring and test if it is a palindrome; track the longest one found.
    // WHEN: Only for very short strings as a correctness baseline.
    public String longestPalindrome_BruteForce(String s) {
        int n = s.length();
        String best = "";
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                if (isPalindrome(s, i, j) && j - i + 1 > best.length()) {
                    best = s.substring(i, j + 1);
                }
            }
        }
        return best;
    }

    private boolean isPalindrome(String s, int l, int r) {
        while (l < r) {
            if (s.charAt(l++) != s.charAt(r--)) return false;
        }
        return true;
    }

    // APPROACH 2: Expand Around Center | O(n²) time | O(1) space
    // EXPLAIN: For each character (and gap between characters), expand outward while chars match.
    // WHEN: Optimal in practice — O(n²) time with O(1) space; standard interview answer.
    public String longestPalindrome_ExpandCenter(String s) {
        int n = s.length();
        int start = 0, maxLen = 1;
        for (int i = 0; i < n; i++) {
            // Odd-length palindromes
            int len1 = expand(s, i, i);
            // Even-length palindromes
            int len2 = expand(s, i, i + 1);
            int len = Math.max(len1, len2);
            if (len > maxLen) {
                maxLen = len;
                start = i - (len - 1) / 2;
            }
        }
        return s.substring(start, start + maxLen);
    }

    private int expand(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }

    // APPROACH 3: Dynamic Programming | O(n²) time | O(n²) space
    // EXPLAIN: dp[i][j] = true if s[i..j] is a palindrome; fill using the recurrence dp[i][j] = (s[i]==s[j]) && dp[i+1][j-1].
    // WHEN: When the interviewer explicitly asks for a DP solution or you need sub-problem reuse.
    public String longestPalindrome_DP(String s) {
        int n = s.length();
        boolean[][] dp = new boolean[n][n];
        int start = 0, maxLen = 1;

        // All substrings of length 1 are palindromes
        for (int i = 0; i < n; i++) dp[i][i] = true;

        // Check substrings of length 2
        for (int i = 0; i < n - 1; i++) {
            if (s.charAt(i) == s.charAt(i + 1)) {
                dp[i][i + 1] = true;
                start = i;
                maxLen = 2;
            }
        }

        // Check substrings of length 3 and above
        for (int len = 3; len <= n; len++) {
            for (int i = 0; i <= n - len; i++) {
                int j = i + len - 1;
                if (s.charAt(i) == s.charAt(j) && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    if (len > maxLen) {
                        maxLen = len;
                        start = i;
                    }
                }
            }
        }
        return s.substring(start, start + maxLen);
    }

    // APPROACH 4: Manacher's Algorithm | O(n) time | O(n) space
    // EXPLAIN: Transform string with '#' separators; reuse palindrome radii via a center-mirror property for linear time.
    // WHEN: When strictly O(n) is required — impressive to know; rarely asked but optimal.
    public String longestPalindrome_Manacher(String s) {
        // Transform: "abc" -> "^#a#b#c#$"
        StringBuilder sb = new StringBuilder("^#");
        for (char c : s.toCharArray()) { sb.append(c); sb.append('#'); }
        sb.append('$');
        String t = sb.toString();
        int n = t.length();
        int[] p = new int[n];
        int C = 0, R = 0;

        for (int i = 1; i < n - 1; i++) {
            int mirror = 2 * C - i;
            if (i < R) p[i] = Math.min(R - i, p[mirror]);
            // Expand around center i
            while (t.charAt(i + p[i] + 1) == t.charAt(i - p[i] - 1)) p[i]++;
            // Update center
            if (i + p[i] > R) { C = i; R = i + p[i]; }
        }

        // Find the index with maximum palindrome radius
        int maxLen = 0, center = 0;
        for (int i = 1; i < n - 1; i++) {
            if (p[i] > maxLen) { maxLen = p[i]; center = i; }
        }
        int start = (center - maxLen) / 2;
        return s.substring(start, start + maxLen);
    }

    // APPROACH 5: DP Space Optimized | O(n²) time | O(n) space
    // EXPLAIN: Use two 1-D boolean arrays (prev, curr) to represent one diagonal of the DP table at a time.
    // WHEN: When you want DP correctness but cannot afford the O(n²) table; trading time constant for memory.
    public String longestPalindrome_DPOptimized(String s) {
        int n = s.length();
        if (n == 0) return "";
        int start = 0, maxLen = 1;
        boolean[] prev = new boolean[n];
        boolean[] cur  = new boolean[n];

        for (int i = n - 1; i >= 0; i--) {
            cur = new boolean[n];
            cur[i] = true;                             // dp[i][i]
            for (int j = i + 1; j < n; j++) {
                if (s.charAt(i) == s.charAt(j)) {
                    cur[j] = (j == i + 1) ? true : prev[j - 1];
                } else {
                    cur[j] = false;
                }
                if (cur[j] && j - i + 1 > maxLen) {
                    maxLen = j - i + 1;
                    start  = i;
                }
            }
            prev = cur;
        }
        return s.substring(start, start + maxLen);
    }
}

// Made with Bob

File: public/python/014-longest-palindromic-substring.py
# LeetCode Problem #5: Longest Palindromic Substring
# Difficulty: Medium
# Link: https://leetcode.com/problems/longest-palindromic-substring/

# ─────────────────────────────────────────────
# APPROACH 1: Brute Force | O(n³) time | O(1) space
# EXPLAIN: Check every substring for palindrome property using a helper; keep the longest found.
# WHEN: Only for tiny inputs — O(n³) is prohibitive even for n ≈ 1000.

def longest_palindrome_brute(s: str) -> str:
    def is_palindrome(l: int, r: int) -> bool:
        while l < r:
            if s[l] != s[r]:
                return False
            l += 1; r -= 1
        return True

    n   = len(s)
    best = s[0] if s else ""
    for i in range(n):
        for j in range(i + 1, n):
            if is_palindrome(i, j) and (j - i + 1) > len(best):
                best = s[i: j + 1]
    return best


# ─────────────────────────────────────────────
# APPROACH 2: Expand Around Center | O(n²) time | O(1) space
# EXPLAIN: For each index (and each pair of adjacent indices) expand outward while characters match.
# WHEN: Clean O(n²) with tiny constant — the most practical solution for interview settings.

def longest_palindrome_expand(s: str) -> str:
    if not s:
        return ""
    start = end = 0

    def expand(l: int, r: int) -> tuple:
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1; r += 1
        return l + 1, r - 1            # last valid palindrome bounds

    for i in range(len(s)):
        l1, r1 = expand(i, i)          # odd length
        l2, r2 = expand(i, i + 1)      # even length
        if r1 - l1 > end - start:
            start, end = l1, r1
        if r2 - l2 > end - start:
            start, end = l2, r2
    return s[start: end + 1]


# ─────────────────────────────────────────────
# APPROACH 3: Dynamic Programming | O(n²) time | O(n²) space
# EXPLAIN: dp[i][j] is True if s[i..j] is a palindrome; fill by length, using the two-character base case.
# WHEN: When you need to enumerate or count all palindromic substrings alongside finding the longest.

def longest_palindrome_dp(s: str) -> str:
    n = len(s)
    if n == 0:
        return ""
    dp = [[False] * n for _ in range(n)]
    start = 0
    max_len = 1
    # All single characters are palindromes
    for i in range(n):
        dp[i][i] = True
    # Check length-2 substrings
    for i in range(n - 1):
        dp[i][i + 1] = (s[i] == s[i + 1])
        if dp[i][i + 1]:
            start   = i
            max_len = 2
    # Check lengths 3..n
    for length in range(3, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1
            dp[i][j] = (s[i] == s[j]) and dp[i + 1][j - 1]
            if dp[i][j] and length > max_len:
                start   = i
                max_len = length
    return s[start: start + max_len]


# ─────────────────────────────────────────────
# APPROACH 4: Manacher's Algorithm | O(n) time | O(n) space
# EXPLAIN: Transform string with separators, then propagate palindrome radius using a center-mirror trick.
# WHEN: When strictly O(n) is required; rarely asked in interviews but impressive to know.

def longest_palindrome_manacher(s: str) -> str:
    # Transform: "abc" -> "#a#b#c#"
    t = "#" + "#".join(s) + "#"
    n = len(t)
    p = [0] * n           # p[i] = radius of palindrome centred at t[i]
    center = right = 0
    for i in range(n):
        if i < right:
            mirror = 2 * center - i
            p[i] = min(right - i, p[mirror])
        # Attempt to expand
        l, r = i - p[i] - 1, i + p[i] + 1
        while l >= 0 and r < n and t[l] == t[r]:
            p[i] += 1
            l -= 1; r += 1
        if i + p[i] > right:
            center, right = i, i + p[i]
    # Find the maximum radius and map back
    max_r = max(p)
    ci    = p.index(max_r)
    start = (ci - max_r) // 2
    return s[start: start + max_r]


# ─────────────────────────────────────────────
# APPROACH 5: DP Space Optimized | O(n²) time | O(n) space
# EXPLAIN: Use two 1-D arrays (prev and curr) instead of the full n×n DP table — one diagonal at a time.
# WHEN: When you want DP correctness but cannot afford O(n²) memory.

def longest_palindrome_dp_optimized(s: str) -> str:
    n = len(s)
    if n == 0:
        return ""
    start = 0
    max_len = 1
    prev = [False] * n
    cur  = [False] * n

    for i in range(n - 1, -1, -1):
        cur = [False] * n
        cur[i] = True                      # dp[i][i] is always True
        for j in range(i + 1, n):
            if s[i] == s[j]:
                cur[j] = True if j == i + 1 else prev[j - 1]
            else:
                cur[j] = False
            if cur[j] and j - i + 1 > max_len:
                max_len = j - i + 1
                start   = i
        prev = cur

    return s[start: start + max_len]


# ─────────────────────────────────────────────
if __name__ == "__main__":
    cases = [
        ("babad",  {"bab", "aba"}),   # two valid answers
        ("cbbd",   {"bb"}),
        ("a",      {"a"}),
        ("racecar", {"racecar"}),
    ]
    for s, valid in cases:
        assert longest_palindrome_brute(s)        in valid
        assert longest_palindrome_expand(s)       in valid
        assert longest_palindrome_dp(s)           in valid
        assert longest_palindrome_manacher(s)     in valid
        assert longest_palindrome_dp_optimized(s) in valid
    print("All tests passed.")

# Made with Bob
